Objective-C/CBLDatabase.mm

```diff
@@ -332,7 +332,23 @@ - (BOOL) inBatch: (NSError**)outError usingBlock: (void (NS_NOESCAPE ^)())block
         if (!transaction.begin())
             return convertError(transaction.error(), outError);
         
-        block();
+        @try {
+            NSError* err = nil;
+            block(&err);
+            if (err) {
+                // if swift throws an exception, `err` will be populated
```

Comment:
Would it be unreasonable just to throw the error, if it in not nil?  Then you wouldn't have to write that code twice.  It isn't really important to make it fast, I think: it is a failure...

Reply:
objc is not designed in a way to handle exceptions. throwing an exceptions will cause a leak, so recommended way is to grab the error and process.

Reply:
If I may....

Swift doesn't have exceptions, what it does have is an error handling mechanism where you can "throw" them and it will return early and they must be handled by the caller with the `do {} catch {}` mechanism or marking the callsite as throwable.

These errors map directly to NSError in Objective C and vice-versa.

Exceptions in Obj-C are complicated but the gist is, you shouldn't catch them in a generic context, but only in a case by case basis where you know it's ok as most NSExceptions are thrown due to unrecoverable errors (out of memory, memory corruption, internal inconsistency, etc) which is why Swift doesn't have them, because well, you can't really do much about them but to abort the program

Reply:
Thank you for the long explanations.  I think you are missing the point.
There is an @catch at line 345, that will catch an NSException raised at line 339.   Doing this would eliminate that annoying duplicate code.

Reply:
Huh? There is no exception being raised at line 339.

If you refer to the comment:
> // if swift throws an exception, `err` will be populated

That is an NSError, which has nothing to do with the Obj-C exceptions mechanism

Reply:
Uhhh...  Oh... NVM... It's committed.

Reply:
I think @dtorres made a good point about the NSException that when it got thrown from the Objective-C code, it's pretty much unrecoverable error.

Reply:
may be the comment is misleading, `if swift throws an error, ..` instead of `throws an exception`. 

@dtorres, even though it says exception, in swift we are catching the `Error` thrown from in..bach block and send back to ObjC as `NSError`(see Swift/Database.swift). Any error happens during the batch operation, this NSError object will be populated.

And yes throwing exception from objc is unrecoverable.

Reply:
Yeah.  Really beating a dead horse, at this point.  I was just pointing out that:
1. That `@catch` catches *something*
2. The code that it executes, when it catches *something* is exactly the same code that is inside the `if` clause
3. Why not `raise` *something*, inside the `if` clause?  Then you don't have to duplicate code.

Reply:
Got your point Blake, but doing so is not recommended in ObjC. 
[reference objc not exceptions-safe](http://clang.llvm.org/docs/AutomaticReferenceCounting.html#exceptions)

The only way we can avoid the duplicate is to write a function, but more overhead. 

Marking this as resolved.

Reply:
I'll be darned!  TIL!  Thank you, @jayahariv !